--- cpp/exporters/otlp/src/otlp_file_client.cc.old	2025-06-20 19:27:44
+++ cpp/exporters/otlp/src/otlp_file_client.cc	2025-06-20 19:27:56
@@ -1253,7 +1253,8 @@
         strerror_s(error_message, sizeof(error_message) - 1, error_code);
 #else
         char error_message[256] = {0};
-        strerror_r(error_code, error_message, sizeof(error_message) - 1);
+        auto ret_ = strerror_r(error_code, error_message, sizeof(error_message) - 1);
+        (void) ret_;
 #endif
         OTEL_INTERNAL_LOG_ERROR("[OTLP FILE Client] Create directory \""
                                 << directory_name << "\" failed.errno: " << error_code
